### PR TITLE
feat: targeting version

### PR DESCRIPTION
**Pull request by gangb-tech**

### CODE REVIEW

File: src/main/java/com/featureprobe/api/service/TargetingService.java [hunk 1]
```diff
@@ -35,18 +44,48 @@ public class TargetingService {
...
+            TargetingVersion targetingVersion = new TargetingVersionBuilder().builder()
+                    .targetingId(targeting.getId())
+                    .projectKey(projectKey)
+                    .environmentKey(environmentKey)
+                    .content(updatedTargeting.getContent())
+                    .version(updatedTargeting.getVersion())
+                    .comment(targetingRequest.getComment())
+                    .build();
+            savaTargetingHistory(targetingVersion);
+        }
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/service/TargetingService.java]: saveTargetingVersion

File: src/main/java/com/featureprobe/api/model/ServerToggleBuilder.java [hunk 2]
```diff
@@ -74,12 +81,11 @@ private Variation.ValueConverter<?> getReturnTypeConverter(String type) {
...
     }
 
-    public Toggle build(String projectKey) {
+    public Toggle build() {
         this.setDisabledServe();
         this.setDefaultServe();
         this.setVariations();
-        this.setRules(projectKey);
-
+        this.setRules(segments);
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/model/ServerToggleBuilder.java]: `segments` is a global variable.

File: src/main/java/com/featureprobe/api/model/ServerToggleBuilder.java [hunk 3]
```diff
@@ -123,7 +129,7 @@ private void setRules(String projectKey) {
         rules.forEach(rule -> rule.getConditions().forEach(condition -> {
             if (condition.getType() != ConditionType.SEGMENT) return;
             condition.setObjects(condition.getObjects().stream().map(segmentKey ->
-                    StringUtils.join(projectKey, '$', segmentKey)).collect(Collectors.toList()));
+                    segments.get(segmentKey).getUniqueKey()).collect(Collectors.toList()));
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/model/ServerToggleBuilder.java]: Extract method: `replaceSegmentKeyToUniqueKey`

File: src/main/java/com/featureprobe/api/repository/SegmentRepository.java [hunk 4]
```diff
@@ -23,4 +25,18 @@ public interface SegmentRepository extends JpaRepository<Segment, Long>, JpaSpec
...
+                    + "OR if(?2 != '', `key` like concat('%',?2,'%'), 1=1) "
+                    + "OR if(?2 != '', description like concat('%',?2,'%'), 1=1) )"
+                    + "AND project_key = ?1 "
+                    + "ORDER BY created_time DESC",
+            countQuery = "SELECT count(*) FROM segment WHERE ( if(?2 != null, name like concat('%',?2,'%'), 1=1) "
+                    + "OR if(?2 != '', `key` like concat('%',?2,'%'), 1=1) "
+                    + "OR if(?2 != '', description like concat('%',?2,'%'), 1=1) )"
+                    + "AND project_key = ?1 ",
+            nativeQuery = true
+    )
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/repository/SegmentRepository.java:39]: Non-standard generic sql, like `OR if`, `concat` .. recommended to use `Specification`

Ref: https://pavankjadda.medium.com/search-data-across-multiple-columns-using-spring-data-jpa-fc0a24799991

File: src/main/java/com/featureprobe/api/service/SegmentService.java [hunk 5]
```diff
@@ -53,9 +53,16 @@ public class SegmentService {
     private EnvironmentRepository environmentRepository;
 
     public Page<SegmentResponse> list(String projectKey, SegmentSearchRequest searchRequest) {
-        Specification<Segment> spec = buildQuerySpec(projectKey, searchRequest.getKeyword());
+        if (searchRequest.getIncludeDeleted()) {
+            Pageable pageable = PageRequest.of(searchRequest.getPageIndex(), searchRequest.getPageSize(),
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/service/SegmentService.java]: PageRequestUtil.toPageable

File: src/main/java/com/featureprobe/api/service/SegmentService.java [hunk 6]
```diff
@@ -53,9 +53,16 @@ public class SegmentService {
...
+                    Sort.Direction.DESC, "created_time");
+            Page<Segment> segments = segmentRepository.findAllByKeywordIncludeDeleted(projectKey,
+                    searchRequest.getKeyword(), pageable);
+            return segments.map(segment -> SegmentMapper.INSTANCE.entityToResponse(segment));
+        }
         Pageable pageable = PageRequest.of(searchRequest.getPageIndex(), searchRequest.getPageSize(),
                 Sort.Direction.DESC, "createdTime");
+        Specification<Segment> spec = buildQuerySpec(projectKey, searchRequest.getKeyword());
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/service/SegmentService.java]: Repeat code.. use Specification?

File: src/main/java/com/featureprobe/api/service/TargetingService.java [hunk 7]
```diff
@@ -35,18 +44,48 @@ public class TargetingService {
...
+        if(updatedTargeting.getVersion() > oldVersion) {
+            TargetingVersion targetingVersion = new TargetingVersionBuilder().builder()
+                    .targetingId(targeting.getId())
+                    .projectKey(projectKey)
+                    .environmentKey(environmentKey)
+                    .content(updatedTargeting.getContent())
+                    .version(updatedTargeting.getVersion())
+                    .comment(targetingRequest.getComment())
+                    .build();
+            savaTargetingHistory(targetingVersion);
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/service/TargetingService.java]: saveTargetingVersion

File: src/main/java/com/featureprobe/api/service/TargetingService.java [hunk 8]
```diff
@@ -35,18 +44,48 @@ public class TargetingService {
...
+        }
+        return TargetingMapper.INSTANCE.entityToResponse(updatedTargeting);
+    }
+
+    private void savaTargetingHistory(TargetingVersion targetingVersion) {
+        targetingVersionRepository.save(targetingVersion);
     }
 
+    public Page<TargetingVersionResponse> versions(String projectKey, String environmentKey,
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/service/TargetingService.java]: Keep naming consistent. `queryVersions`

File: src/main/java/com/featureprobe/api/service/TargetingService.java [hunk 9]
```diff
@@ -35,18 +44,48 @@ public class TargetingService {
...
+                                                   TargetingVersionRequest targetingVersionRequest) {
+        Pageable pageable = PageRequest.of(targetingVersionRequest.getPageIndex(),
+                targetingVersionRequest.getPageSize(), Sort.Direction.DESC, "createdTime");
+        Page<TargetingVersion> targetingVersions = targetingVersionRepository
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/service/TargetingService.java:83]: Used`PageRequestUtil.toPageable`  or add method: `PageRequestUtil.toCreatedTimeDescSortPageable(targetingVersionRequest)`

File: src/main/java/com/featureprobe/api/service/TargetingService.java [hunk 10]
```diff
@@ -90,4 +129,49 @@ private void validateTargetingContent(String projectKey, TargetingContent conten
 
     }
 
+    class TargetingVersionBuilder {
```
Comment (dengfeige) [inline src/main/java/com/featureprobe/api/service/TargetingService.java]: `builder 模式` 不是用来代替 `set` 的。这里其实不应该用 builder

**Review by dengfeige — CHANGES_REQUESTED** [review]

File: src/main/java/com/featureprobe/api/repository/SegmentRepository.java (hunk 4 above)
Comment (gangb-tech) [reply src/main/java/com/featureprobe/api/repository/SegmentRepository.java:39]: entity used @Where(clause = "deleted = 0") ，so need native Query

**Pull request merged**